File: examples/claude-agent-sdk-ts/src/tools.ts
// Claude Agent SDK with an in-process custom tool, routed through
// VM-X. `createSdkMcpServer` exposes our tool to the agent loop via
// an embedded MCP server; the model's tool calls + tool results flow
// through transparently.

import {
  createSdkMcpServer,
  query,
  tool,
} from '@anthropic-ai/claude-agent-sdk';
import { z } from 'zod';
import { load, optionsEnv } from './config.js';

const getWeather = tool(
  'get_weather',
  'Get the current weather for a city.',
  { city: z.string().describe('City name') },
  async ({ city }) => {
    let text: string;
    const lc = city.toLowerCase();
    if (lc.includes('são paulo') || lc.includes('sao paulo')) {
      text = 'São Paulo: 22°C, cloudy';
    } else if (lc.includes('rio')) {
      text = 'Rio de Janeiro: 28°C, sunny';
    } else {
      text = `${city}: temperature unknown`;
    }
    return { content: [{ type: 'text', text }] };
  }
);

async function main(): Promise<void> {
  const cfg = load();
  const server = createSdkMcpServer({
    name: 'weather',
    version: '1.0.0',
    tools: [getWeather],
  });

  const prompt =
    'What is the weather in São Paulo and Rio de Janeiro? ' +
    'Use the get_weather tool for each city.';

  for await (const message of query({
    prompt,
    options: {
      env: optionsEnv(cfg),
      model: cfg.anthropicResource,
      mcpServers: { weather: server },
      allowedTools: ['mcp__weather__get_weather'],
      ...(cfg.pathToClaudeCodeExecutable
        ? { pathToClaudeCodeExecutable: cfg.pathToClaudeCodeExecutable }
        : {}),
    },
  })) {
    if (message.type === 'assistant') {
      for (const block of message.message.content) {
        if (block.type === 'text') {
          console.log(`[assistant text] -> ${block.text}`);
        } else if (block.type === 'tool_use') {
          console.log(
            `[assistant tool_use] -> ${block.name}(${JSON.stringify(
              block.input
            )})`
          );
        }
      }
    }
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/claude-agent-sdk-python/claude_agent_sdk_example/tools.py
"""Claude Agent SDK with an in-process custom tool, routed through VM-X.

The agent calls a Python function via the in-process MCP server the SDK
creates for us; the model's tool calls + tool results flow through the
agentic loop transparently.

Run with:

    pnpm exec nx run claude-agent-sdk-python-example:tools
"""

from __future__ import annotations

import asyncio

from claude_agent_sdk import (
    ClaudeAgentOptions,
    create_sdk_mcp_server,
    query,
    tool,
)

from .config import load, options_env


@tool("get_weather", "Get the current weather for a city", {"city": str})
async def get_weather(args: dict) -> dict:
    city = (args.get("city") or "").lower()
    if "são paulo" in city or "sao paulo" in city:
        result = "São Paulo: 22°C, cloudy"
    elif "rio" in city:
        result = "Rio de Janeiro: 28°C, sunny"
    else:
        result = f"{args.get('city', 'unknown')}: temperature unknown"
    return {"content": [{"type": "text", "text": result}]}


async def main() -> None:
    cfg = load()
    server = create_sdk_mcp_server(
        name="weather", version="1.0.0", tools=[get_weather]
    )

    options = ClaudeAgentOptions(
        env=options_env(cfg),
        model=cfg.anthropic_resource,
        mcp_servers={"weather": server},
        allowed_tools=["mcp__weather__get_weather"],
    )

    prompt = (
        "What is the weather in São Paulo and Rio de Janeiro? "
        "Use the get_weather tool for each city."
    )
    async for message in query(prompt=prompt, options=options):
        msg_type = type(message).__name__
        content = getattr(message, "content", None) or getattr(
            getattr(message, "message", None), "content", None
        )
        if content is None or not isinstance(content, list):
            continue
        for block in content:
            text = getattr(block, "text", None)
            if text:
                print(f"[{msg_type}] -> {text}")
                continue
            name = getattr(block, "name", None)
            inp = getattr(block, "input", None)
            if name is not None:
                print(f"[{msg_type}] tool_use -> {name}({inp!r})")


if __name__ == "__main__":
    asyncio.run(main())
